File: leetcode/solutions/0118.PascalsTriangle/generate.ts
export default function generate(numRows: number): number[][] {
  if (numRows === 0) {
    return [];
  }
  if (numRows === 1) {
    return [[1]];
  }
  if (numRows === 2) {
    return [[1], [1, 1]];
  }

  const result: number[][] = [[1], [1, 1]];

  for (let i = 2; i < numRows; i++) {
    let currentRow = [1];
    currentRow[i] = 1;
    const lastRow = result[i - 1];
    const max = Math.floor(i / 2);

    for (let j = 1; j <= max; j++) {
      if (j === max && i % 2 === 0) {
        currentRow[j] = lastRow[j - 1] * 2;
        break;
      }

      currentRow[j] = currentRow[i - j] = lastRow[j - 1] + lastRow[j];
    }

    result.push(currentRow);
  }

  return result;
}

File: leetcode/solutions/0118.PascalsTriangle/generate.go
package solution

func generate(numRows int) [][]int {
	if numRows == 0 {
		return [][]int{}
	}
	if numRows == 1 {
		return [][]int{{1}}
	}
	if numRows == 0 {
		return [][]int{{1}, {1, 1}}
	}

	result := make([][]int, numRows-1)
	result[0] = []int{1}
	result[1] = []int{1, 1}

	for i := 2; i <= numRows; i++ {
		result[i] = make([]int, i+1)
		result[i][0] = 1
		result[i][i] = 1
		last := numRows / 2

		for j := 1; j <= last; j++ {
			if j == last {
				result[i][j] = result[i-1][j-1] * 2
			} else {
				result[i][j] = result[i-1][j-1] + result[i-1][j]
				result[i][i-j] = result[i][j]
			}
		}
	}

	return result
}

func reverse(arr []int) []int {
	last := len(arr) / 2

	for left := 0; left <= last; left++ {
		right := len(arr) - 1 - left
		arr[left], arr[right] = arr[right], arr[left]
	}

	return arr
}
